core/src/apps/cardano/sign_tx/signer.py

```diff
@@ -749,6 +1005,115 @@ def _validate_required_signer(
+            await self._show_collateral_return_init(output)
+
+        # We allow only collateral return outputs controlled by a key, see
+        # _validate_collateral_return. This implies no datum or reference script.
+        output_items_count = 2
+        if output.format == CardanoTxOutputSerializationFormat.LEGACY:
+            output_list: HashBuilderList = HashBuilderList(output_items_count)
+            with self.tx_dict.add(TX_BODY_KEY_COLLATERAL_RETURN, output_list):
+                await self._process_legacy_output(
+                    output_list, output, should_show_tokens
+                )
+        elif output.format == CardanoTxOutputSerializationFormat.POST_ALONZO:
+            output_dict: HashBuilderDict[int, Any] = HashBuilderDict(
+                output_items_count, wire.ProcessError("Invalid collateral return")
+            )
+            with self.tx_dict.add(TX_BODY_KEY_COLLATERAL_RETURN, output_dict):
+                await self._process_post_alonzo_output(
+                    output_dict, output, should_show_tokens
+                )
+        else:
+            raise RuntimeError  # should be unreachable
+
+    def _validate_collateral_return(self, output: messages.CardanoTxOutput) -> None:
+        self._validate_output(output)
+
+        address_type = self._get_output_address_type(output)
+        if address_type not in addresses.ADDRESS_TYPES_PAYMENT_KEY:
+            wire.ProcessError("Invalid collateral return")
+
+    async def _show_collateral_return_init(
+        self, output: messages.CardanoTxOutput
+    ) -> None:
+        # We don't display missing datum warning since only non-script addresses
+        # are allowed.
+
+        if output.asset_groups_count > 0:
+            await layout.warn_tx_output_contains_tokens(self.ctx)
+
+        if output.address_parameters is not None:
+            address = addresses.derive_human_readable(
+                self.keychain,
+                output.address_parameters,
+                self.msg.protocol_magic,
+                self.msg.network_id,
+            )
+            await self._show_output_credentials(
+                output.address_parameters,
+            )
+        else:
+            assert output.address is not None  # _validate_output
+            address = output.address
+
+        await layout.confirm_sending(
+            self.ctx,
+            output.amount,
+            address,
+            "Collateral return",
+            self.msg.network_id,
+        )
+
+    def _should_show_collateral_return_init(
+        self, output: messages.CardanoTxOutput
+    ) -> bool:
+        if self.msg.total_collateral is None:
+            return True
+
+        if self._is_simple_change_output(output):
+            return False
+
+        return True
+
+    def _should_show_collateral_return_tokens(
+        self, output: messages.CardanoTxOutput
+    ) -> bool:
+        return not self._is_simple_change_output(output) and self.is_expert_view
```

Comment:
I don't like the asymmetry between this and the corresponding condition in `_should_show_collateral_return_init` - what do you think about keeping the more verbose wording in order to make the similarity with `_should_show_collateral_return_init` more obvious?

```python
def _should_show_collateral_return_tokens(
        self, output: messages.CardanoTxOutput
    ) -> bool:
        if self._is_simple_change_output(output):
            return False

        return self.is_expert_view
```

Reply:
https://github.com/vacuumlabs/trezor-firmware/pull/79/commits/4b12f34980e074fe2bf395964d1293198fbe1202